### ISSUE TITLE
Bug in WGAN-GP ?

### ISSUE THREAD

**Issue by pfeatherstone** [eriklindernoren/PyTorch-GAN#136]
Should the following https://github.com/eriklindernoren/PyTorch-GAN/blob/a163b82beff3d01688d8315a3fd39080400e7c01/implementations/wgan_gp/wgan_gp.py#L162  be this instead:
```
fake_imgs = generator(z).detach()
``` 
?

**Comment by vegetable09** [comment]
![图片](https://user-images.githubusercontent.com/15920086/100510473-41e9af80-31a9-11eb-85c9-9a5b155562ed.png)

the argorithm in the wgan-gp paper is above. 
There are two ways to calulate G and D iteratively in pytorch:
1. 
```
# update D
fake = G(z)
fake.detach()  # add detach() so the backward() will NOT change the weights of G
loss_d = D(fake) - D(real)
loss_d.backward()
optimizerD.step()  # update the weights of D

# then update G
loss_g = -D(fake)
loss_g.backward()
optimizerG.step()  # update the weights of G
```
2.
```
# update D
fake = G(z)
# no detach() here so the backward() will change the weights of G
loss_d = D(fake) - D(real)
loss_d.backward(retain_graph=True)  # Otherwise, the graph will be delete so the loss_g.backward() will failed
optimizerD.step()  # update the weights of D

# then update G
loss_g = -D(fake)
loss_g.backward()
optimizerG.step()  # update the weights of G
```

**Comment by PerryXDeng** [comment]
Doesn't seem to be necessary for what the code is trying to do, as loss won't propagate beyond `fake` for the purpose of training the discriminator. So while it might not be memory efficient or a 100% faithful reimplementation of the paper, it won't introduce bugs.

**Comment by annan-tang** [comment]
Hi, it may be too late to answering this question. In my opinion, it isn't a bug but can slow down the computation speed. In the original code of wgan-gp, it use separate optimizers for G and D. When you are training the D by use the fake_image from G without detach this data(tensor), the loss_d.backward() will calculate the gradients for both D and G parts( even G part isn't necessary in D training stage), but in D training stage, the source code only update the parameters of D by   **optimizerD.step()** and at the beginning of G training stage, the source code first clear the gradients calculated from D training stage with **zero_grad()** .
so it's not a bug, but add additional computational overhead on D training. You can add the .detach() to make your code more efficient!
